### ISSUE TITLE
BUG: CNNPatchPredictor Predict Merge Outputs Produces ValueError

### ISSUE THREAD

**Issue by John-P** [TissueImageAnalytics/tiatoolbox#186]
* TIA Toolbox version: 0.8.0
* Python version: 3.7.8
* Operating System: Pop!_OS 21.04 (Ubuntu 21.04)

### Description

Using `CNNPatchPredictor.predict` with the `merge_predictions` argument set to True can produce a `ValueError` during merging after all patches are predicted. Setting `merge_predictions` to False does not cause an exception.

### What I Did

Try to run `CNNPatchPredictor.predict` with the `merge_predictions` argument set true. Example code:

```python
from tiatoolbox.wsicore.wsireader import WSIReader
from tiatoolbox.models.controller import patch_predictor

wsi = WSIReader.open("./TCGA-A6-5662-01Z-00-DX1.82569684-1c31-4346-af9b-c296a020f624.svs")
predictor = patch_predictor.CNNPatchPredictor(pretrained_model="resnet18-kather100k")
output = predictor.predict([wsi.input_path], mode="wsi", merge_predictions=True)
```

Error output:

```
100%|#########################################| 378/378 [00:56<00:00,  6.65it/s]
Traceback (most recent call last):
  File ".../issue.py", line 19, in <module>
  File ".../tiatoolbox/models/controller/patch_predictor.py", line 531, in predict
    postproc_func=self.model.postproc,
  File ".../tiatoolbox/models/controller/patch_predictor.py", line 216, in merge_predictions
    canvas_shape = reader.slide_dimensions(resolution=resolution, units=units)
  File ".../tiatoolbox/wsicore/wsireader.py", line 509, in slide_dimensions
    [0, 0] + list(wsi_shape_at_baseline), resolution, units, precisions
  File ".../tiatoolbox/wsicore/wsireader.py", line 545, in _find_read_bounds_params
    resolution, units, precision
  File ".../tiatoolbox/wsicore/wsireader.py", line 285, in _find_optimal_level_and_downsample
    level_scales = self._relative_level_scales(resolution, units)
  File ".../tiatoolbox/wsicore/wsireader.py", line 253, in _relative_level_scales
    raise ValueError(f"Invalid units `{units}`.")
ValueError: Invalid units `None`.
```

**Comment by John-P** [comment]
I also noticed that passing in a WSIReader object raised and exception and only passing in ten file path worked. I think that a WSIReader object should also work here.

**Comment by vqdang** [comment]
Addressed in #188

**Comment by vqdang** [comment]
> 
> 
> I also noticed that passing in a WSIReader object raised and exception and only passing in the file path worked. I think that a WSIReader object should also work here.

We discussed long ago to only use input as list of paths though. Well it was to limit the format checking. If we want to extend to full list of paths, list of nd, list of wsireader etc. will need to fully write a protocol for format checking. At the moment they are all duplicated here and there.
